source/common/buffer/watermark_buffer.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/buffer/buffer.h"
+
+namespace Envoy {
+namespace Buffer {
+
+// A wrapper for an underlying buffer which does watermark validation.
+// The underlying buffer's ownership is transfered to the Watermark buffer.   Each time the inner
+// buffer is resized (written to or drained), the watermarks are checked.  As the buffer size
+// transitions from under the low watermark to above the high watermark, the above_high_watermark
+// function is called one time. It will not be called again until the buffer is drained below the
+// low watermark, at which point the below_low_watermark function is called.
+class WatermarkBuffer : public Instance {
```

Comment:
Sticking with things as is SGTM.  
Having thought more I agree with Harvey that this is more of a code cleanliness choice rather than a perf issue so I'd lean towards adding a TODO iff it shows up in profiles if that's OK